### DISCUSSION TITLE
How to customize request and response that is getting captured

### DISCUSSION THREAD

**Discussion by akkgit** [thepirat000/Audit.NET#698] [General]
I am looking to mask some of the fields in request, response for not to store in logs such as PII, In this scenario how to customize it ? and also the log it is going to save using data providers doesnt block the main thread correct(is it kind of fire and forget) ?

**Comment by thepirat000** [comment]
Could you please provide further details about the package extensions and the framework you're utilizing?

And how and where do you intend to store the audit logs?

I assume you're using ASP.NET Core with `Audit.WebApi.Core` extension and you want a DataProvider that operates in a fire-and-forget manner, but it will be easier to help if you provide enough information.

**Comment by thepirat000** [comment]
Regarding the fire-and-forget mechanism, the data provider's save operation is intentionally blocking by design. However, you can easily implement a custom data provider by inheriting from the data provider of your choice and overriding the Insert/Replace methods to make fire-and-forget calls to the base methods.

This type of data provider is not included in the library because there are numerous ways to implement it, and personally, I believe it's generally a poor practice.

You also have the option to use an in-memory data provider to delegate data storage operations to a separate thread.

- [In-Memory Blocking Collection](https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.NET/Providers/BlockingCollectionDataProvider.cs): Store the audit events in a BlockingCollection that can be accessed by different threads to consume the events
- [In-Memory Channel](https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.NET.Channels/README.md): Store the audit events in a Channel (from System.Threading.Channels) that can be accessed to consume the events
